### PR TITLE
Improve signature help by more stable position calculation + better named arg support

### PR DESCRIPTION
**Pull request by rochala**
Previously, signature help had unstable way of calculating active parameter inside the signature help. It is now changed to work better with erroneous trees such as unclosed openings.

It also adds a new feature for signature help, which will help user navigate inside named arguments. Furthermore, it will now find reordered named arguments, and display in the same way, by also marking the remaining parameters that they are now required to be named.

Example:

https://github.com/lampepfl/dotty/assets/48657087/b181d2d5-60f0-46a5-b2df-a58aa5f07454

For the reviewers: I recommend testing it locally.

### CODE REVIEW

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala [hunk 1]
```diff
@@ -1,79 +1,134 @@
...
+   *  To support getting the tree withh unclosed opening, when our cursor is on the newline such as:
+   *  ```scala
+   *  def foo(aaa: Int, bbb: Int) = ???
+   *  foo(1,
+   *
+   *  @@
+   *  // eof or illegal token
+   *  ```
+   *  we need to rely on dianostics to locate last tree before the cursor
+   *  and `opnning closed` expected syntax error.
```
Comment (kasiaMarek) [inline presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:81]:
```suggestion
   *  and `opening closed` expected syntax error.
```

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala [hunk 2]
```diff
@@ -1,79 +1,134 @@
...
+   *
+   *  To support getting the tree withh unclosed opening, when our cursor is on the newline such as:
+   *  ```scala
+   *  def foo(aaa: Int, bbb: Int) = ???
+   *  foo(1,
+   *
+   *  @@
+   *  // eof or illegal token
+   *  ```
+   *  we need to rely on dianostics to locate last tree before the cursor
```
Comment (kasiaMarek) [inline presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:80]:
```suggestion
   *  we need to rely on diagnostics to locate last tree before the cursor
```

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala [hunk 3]
```diff
@@ -1,79 +1,134 @@
...
-    val (paramN, callableN, alternatives) =
-      Signatures.signatureHelp(path, pos.span)
-    val infos = alternatives.flatMap { signature =>
-      signature.denot.map {
-        (signature, _)
+  /** Interactive.pathTo returns a path to the tree, by relying on Span.contains(otherSpan).
+   *  When there is an unclosed opening, parser will try to repair it, but the span of that tree
+   *  will end on the new line or next illegal token.
+   *
+   *  To support getting the tree withh unclosed opening, when our cursor is on the newline such as:
```
Comment (kasiaMarek) [inline presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:72]:
```suggestion
   *  To support getting the tree with unclosed opening, when our cursor is on the newline such as:
```

**Review by kasiaMarek — COMMENTED** [review]
Also the unclosed opening works for params but not type params.

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala [hunk 4]
```diff
@@ -1,79 +1,134 @@
...
+    val sourceFile = SourceFile.virtual(uri, text)
     driver.run(uri.nn, sourceFile)
 
-    given ctx: Context = driver.currentCtx
-
-    val pos = driver.sourcePosition(params)
-    val trees = driver.openedTrees(uri.nn)
-
-    val path = Interactive.pathTo(trees, pos)
+    driver.compilationUnits.get(uri) match
```
Comment (jkciesluk) [inline presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:37]: Why this change?

**Review by jkciesluk — CHANGES_REQUESTED** [review]
Great job! It fixes a lot of things, but I'm unsure if handling unclosed openings is needed, especially because it really complicates finding enclosing apply.

Could you add these tests:
```scala
def test[T, F](aaa: Int, bbb: T, ccc: F): T = ???
val x = test(1, ccc = 2, b@@)
```

```scala
def test2(aaa: Int, bbb: Int, ccc: Int): Int = ???
val x = test2(
    1, 
    @@
  )
```

```scala
def test2(aaa: Int, bbb: Int, ccc: Int): Int = ???
val x = test2(aaa = 2@@, ccc = 1)
```

```scala
def test2(aaa: Int, bbb: Int, ccc: Int): Int = ???
val x = test2(aaa = 2, @@  ccc = 1, bbb = 3)
```

Also, would it be possible to show in the case below that arguments in the first parameter list are reordered?

```scala
def test3(aaa: Int, bbb: Int, ccc: Int)(ddd: Int): Int = ???
val x = test3(bbb = 1, aaa = 2, ccc = 3)(d@@)
```

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala (hunk 4 above)
Comment (rochala) [reply presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:37]: We need untpd trees, and without this you don't have this information in context.

**Comment by rochala** [comment]
The following changes have been made:
- we will stop supporting Signature Helps for unclosed openings. It is too hard to recover from such errors reliably and to track where we are with the cursor. All modern editors will close opening on input. It simplified the code, and is the best way forward.
- if there is an error within a definition return type parameter, we will now fall back to source, instead of print error,
- tuples and functions are now supported when they are returned from methods.
- this rewrite made me change a few things, and I took the chance and made it support clause interleaving. There is now a proper test suite for that. It required changing the way we represent Signatures.
- argument reordering now works for all parameter lists, not only the currently applied one,
- type parameters now also correctly display documentation

**Comment by rochala** [comment]
I've changed the PR to use `ShortenedTypePrinter` and fixed some other minor issues like correctly marking reordered argumentss or hiding synthetic parameter names. 

In the future, Signature help should be changed to return symbols from the compiler instead of already printed values, and only print them later using shared API of ShortenedTypePrinter in the Presentation Compiler. Ideally we'd want a shared implementation for label printing with completions, hovers and other features.

**Review by jkciesluk — APPROVED** [review]
LGTM! Great work!

File: compiler/src/dotty/tools/dotc/util/Signatures.scala [hunk 5]
```diff
@@ -397,85 +452,136 @@ object Signatures {
...
+        match
+          case Some(evidenceTypeName) => TypeParam(s"${name.show}: ${evidenceTypeName}")
+          case None => TypeParam(name.show + info.show)
+
+    def toParamss(tp: Type, fun: Option[untpd.GenericApply])(using Context): List[List[Param]] =
+      val paramSymss = symbol.paramSymss
+
+      def reduceToParamss(applies: List[untpd.Tree], types: List[Type], paramList: Int = 0): List[List[Param]] =
+        applies -> types match
+          case ((_: untpd.TypeApply) :: restTrees, (poly: PolyType) :: restTypes) =>
```
[the remark this answers is not in the document]
Comment (rochala) [reply compiler/src/dotty/tools/dotc/util/Signatures.scala:507]: It matches against non empty apply list. Removing it failed the test.

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala [hunk 6]
```diff
@@ -1,134 +1,149 @@
...
+
+        val pos = driver.sourcePosition(params, isZeroExtent = false)
+        val path = Interactive.pathTo(unit.tpdTree, pos.span)(using driver.currentCtx)
+
+        val localizedContext = Interactive.contextOfPath(path)(using driver.currentCtx)
+        val indexedContext = IndexedContext(driver.currentCtx)
+
+        given Context = localizedContext.fresh
+          .setCompilationUnit(unit)
+          .setPrinterFn(_ => ShortenedTypePrinter(search, IncludeDefaultParam.Include)(using indexedContext))
```
Comment (rochala) [inline presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:51]: This should be not include

File: presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpNamedArgsSuite.scala [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+      """|object O:
+         |  def method(paramA: Int, paramB: Int, paramC: Int, paramD: Int): Unit = ???
+         |  method(paramA = 1, @@)
+         |""".stripMargin,
+      """|method(paramA: Int, paramB: Int, paramC: Int, paramD: Int): Unit
+         |                    ^^^^^^^^^^^
+         |""".stripMargin
+    )
+
+  @Test def `new-named-param-style-double-space` =
```
Comment (kasiaMarek) [inline presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpNamedArgsSuite.scala:21]: Unclosed opening test?

File: language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala [hunk 8]
```diff
@@ -930,23 +935,25 @@ object DottyLanguageServer {
...
-      else
-        signature.paramss
-          .map { paramList =>
-            val labels = paramList.map(_.show)
-            val prefix = if paramList.exists(_.isImplicit) then "using " else ""
-            labels.mkString(prefix, ", ", "")
-          }
-          .mkString("(", ")(", ")")
-    val tparamsLabel = if (signature.tparams.isEmpty) "" else signature.tparams.mkString("[", ", ", "]")
+      signature.paramss
```
Comment (kasiaMarek) [inline language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:941]: I assume this whole duplication is because there is no good place for it, such that it could be used by both modules.

File: presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala [hunk 9]
```diff
@@ -1,134 +1,149 @@
...
-     */
-    new l.SignatureHelp(
-      signatureInfos.map(signatureToSignatureInformation).asJava,
-      callableN,
-      paramN
-    )
+    driver.compilationUnits.get(uri) match
+      case Some(unit) =>
+
+        val pos = driver.sourcePosition(params, isZeroExtent = false)
```
Comment (kasiaMarek) [inline presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:43]: If you're expanding offset to a range span, won't that throw an error later if the cursor is at the very end of the file? I know, we don't do open brackets but we should make sure not to throw.

**Review by kasiaMarek — APPROVED** [review]
Looks really good. 🚀  I feel like you improved naming of different variables/methods, thanks for that, I think it made things more readable.

File: presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpPatternSuite.scala [hunk 10]
```diff
@@ -202,23 +202,6 @@ class SignatureHelpPatternSuite extends BaseSignatureHelpSuite:
          |""".stripMargin
     )
 
-  @Test def `pat4` =
```
[the remark this answers is not in the document]
Comment (rochala) [reply presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpPatternSuite.scala:205]: A. It is not working any more due to changes to range.
B. It was a bit useless as it would not be automatically opened and triggered, also the information is not that helpful, especially if you have more than single Apply like
`a :: b :: c :: Nil`

File: presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpSuite.scala [hunk 11]
```diff
@@ -323,9 +323,9 @@ class SignatureHelpSuite extends BaseSignatureHelpSuite:
         |  } yield k
         |}
       """.stripMargin,
-      """|to(end: Int): scala.collection.immutable.Range.Inclusive
+      """|to(end: Int): Inclusive
```
[the remark this answers is not in the document]
Comment (rochala) [reply presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpSuite.scala:326]: No, it should only shorten packages. I'll change it in separate PR as this is not caused by these changes.

File: language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala (hunk 8 above)
Comment (rochala) [reply language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:941]: Dotty language server should actually be removed from the repository as it is no longer necessary. It is not so easy tho as the test coverage is different. This change was only made to make the tests pass
